start.py

```diff
@@ -1,47 +1,230 @@
 
+class Player:
+    def __init__(self, name):
+        self._name = name
+        self._board = None
 
-def print_available_pieces(pieces):
-    print("Available Pieces:")
-    sorted_pieces = sorted(pieces)
-    for piece in sorted_pieces:
-        print(piece, end=" ")
+    def join_board(self, board):
+        self._board = board
 
+    def place(self, piece):
+        print("Otrzymana figura: \n%s" % piece)
+        while True:
+            row = input("Wybierz rząd (1-4): ")
+            while not (row.isnumeric() and 1 <= int(row) <= 4):
+                row = input("Wybrany rząd nie istnieje, ponów wybór: ")
+            row = int(row)
+            
+            col = input("Wybierz kolumnę (1-4): ")
+            while not (col.isnumeric() and 1 <= int(col) <= 4):
+                col = input("Wybrana kolumna nie istnieje, ponów wybór: ")
+            col = int(col)
+            
+            if self._board.is_avaliable(row, col):
+                self._board.place(piece, row, col)
+                quarto_call = input("Czy chcesz zawołać quatro? Wybierz TAK/NIE [t/n]: ")
+                while not quarto_call.upper() in ("T", "TAK", "N", "NIE"):
+                    quarto_call = input("Odpowiedź musi być w formacie TAK lub NIE [t/n]: ")
+                if quarto_call.upper() in ("T", "TAK"):
+                    if self._board.is_winning(piece):
+                        return True
+                    print("Podana figura nie tworzy Quatro")
+                return False
+            
+            print("Wybrane pole jest zajęte, wybierz inne.")
 
-def is_quarto(board):
-    return False
+    def __str__(self):
+        return self._name
+
+class Piece:
+    def __init__(self, piece_value):
+        if piece_value == None or piece_value < 0 or 16 <= piece_value: # empty piece case
+            self._value = -1
+            self._bin_value = None
+            return
+        
+        self._value = piece_value
+
+        self._bin_value = [0, 0, 0, 0]
+        if piece_value >= 8: # color
+            self._bin_value[0] = 1
+            piece_value -= 8
+        if piece_value >= 4: # shape
+            self._bin_value[1] = 1
+            piece_value -= 4
+        if piece_value >= 2: # filling
+            self._bin_value[2] = 1
+            piece_value -= 2
+        self._bin_value[3] = piece_value # height
+
+    def is_empty(self):
+        return self._bin_value == None
+
+    def is_white(self):
+        return self._bin_value[0] == 1
+
+    def is_circle(self):
+        return self._bin_value[1] == 1
+
+    def is_square(self):
+        return self._bin_value[2] == 1
+
+    def is_high(self):
+        return self._bin_value[3] == 1
+
+    def decimal(self): # empty pieces return -1
+        return self._value
+
+    def binary(self): # empty pieces return None
+        return self._bin_value
+
+    def __str__(self):
+        return "%d%d\n%d%d" % (self._bin_value[0], self._bin_value[1], self._bin_value[2], self._bin_value[3])
```

Comment:
proponuję korzystać z f string. dla tego przypadku to byłoby:

+ return f"{self._bin_value[0]}{self._bin_value[1]}\n{self._bin_value[2]}{self._bin_value[3]}"

Reply:
Stosowne usprawnienie.